x-pack/plugins/security_solution/public/detection_engine/rule_management/api/hooks/use_update_rule_mutation.ts

```diff
@@ -6,42 +6,43 @@
  */
 import type { UseMutationOptions } from '@tanstack/react-query';
 import { useMutation } from '@tanstack/react-query';
-import type {
-  RuleResponse,
-  RuleUpdateProps,
-} from '../../../../../common/api/detection_engine/model/rule_schema';
+import type { RuleUpdateProps } from '../../../../../common/api/detection_engine/model/rule_schema';
 import { transformOutput } from '../../../../detections/containers/detection_engine/rules/transforms';
 import { DETECTION_ENGINE_RULES_URL } from '../../../../../common/constants';
 import { updateRule } from '../api';
 import { useInvalidateFindRulesQuery } from './use_find_rules_query';
-import { useInvalidateFetchRuleByIdQuery } from './use_fetch_rule_by_id_query';
+import { useUpdateRuleByIdCache } from './use_fetch_rule_by_id_query';
 import { useInvalidateFetchRuleManagementFiltersQuery } from './use_fetch_rule_management_filters_query';
 import { useInvalidateFetchCoverageOverviewQuery } from './use_fetch_coverage_overview';
+import type { Rule } from '../../logic/types';
 
 export const UPDATE_RULE_MUTATION_KEY = ['PUT', DETECTION_ENGINE_RULES_URL];
 
 export const useUpdateRuleMutation = (
-  options?: UseMutationOptions<RuleResponse, Error, RuleUpdateProps>
+  options?: UseMutationOptions<Rule, Error, RuleUpdateProps>
 ) => {
   const invalidateFindRulesQuery = useInvalidateFindRulesQuery();
   const invalidateFetchRuleManagementFilters = useInvalidateFetchRuleManagementFiltersQuery();
-  const invalidateFetchRuleByIdQuery = useInvalidateFetchRuleByIdQuery();
   const invalidateFetchCoverageOverviewQuery = useInvalidateFetchCoverageOverviewQuery();
+  const updateRuleCache = useUpdateRuleByIdCache();
 
-  return useMutation<RuleResponse, Error, RuleUpdateProps>(
+  return useMutation<Rule, Error, RuleUpdateProps>(
     (rule: RuleUpdateProps) => updateRule({ rule: transformOutput(rule) }),
     {
       ...options,
       mutationKey: UPDATE_RULE_MUTATION_KEY,
       onSettled: (...args) => {
         invalidateFindRulesQuery();
-        invalidateFetchRuleByIdQuery();
         invalidateFetchRuleManagementFilters();
         invalidateFetchCoverageOverviewQuery();
 
-        if (options?.onSettled) {
-          options.onSettled(...args);
+        const [response] = args;
+
+        if (response) {
+          updateRuleCache(response);
         }
+
+        options?.onSettled?.(...args);
```

Comment:
Nit: while optional chaining simplifies accessing object attributes, using it with function calls can reduce code readability. It becomes more difficult to spot the various code paths being executed given different inputs. And one needs to read through the code more carefully to understand it:

```ts
if (foo?.bar?.baz) // <- a short glance is enough to notice this is an optional call
  foo.bar.baz(args); 
}
```

```ts
foo?.bar?.baz?.(args); // <- really need to read this line carefully to spot the optional call
```

 So I would advise against using optional chaining for function calls.

Reply:
Tbh I'm not 100% agree that the first option is always better. Having too many `if` statements make the code bulky and less readable. The common sense should be a driver here. In this particular case I decided  `options?.onSettled?.(...args);` is readable enough while it can be less obvious for someone unfamiliar with optional chaining and it can be applied to function calls.